content/doc/book/pipeline/development.adoc

```diff
@@ -0,0 +1,145 @@
+---
+layout: section
+title: Pipeline Development Tools
+---
+:description:
+:author:
+:email: jenkinsci-docs@googlegroups.com
+:sectanchors:
+:toc:
+:imagesdir: /doc/book/resources
+:hide-uri-scheme:
+
+= Pipeline Development Tools
+
+In this section, we'll discuss tools and resources
+that help speed development of Jenkins Pipelines.
+
+== Blue Ocean Editor (Declarative only)
+
+The
+<<../blueocean/pipeline-editor#, Blue Ocean Pipeline Editor>> provides a WYSIWYG
+way to create Declarative Pipelines. It gives you a structural view of all the stages,
+parallel branches, and steps in your Pipeline. It validates Pipeline changes as they are
+made, eliminating many errors before they are even committed.  Behind the scenes
+it still generates Declarative Pipeline code.
+
+[[linter]]
+== Command-line Pipeline Linter (Declarative only)
+
+Jenkins can validate, or
+"link:https://en.wikipedia.org/wiki/Lint_(software)[lint]",
+a Declarative Pipleline from the command line before running actually running it.
+This can be done using a Jenkins CLI command or using `curl` to post the Jenkinsfile directly.
+We recommended using the
+<<../managing/cli#ssh, SSH interface>>
+for running the linter.
+
+NOTE: See the <<../managing/cli, Jenkins CLI documentation>> for details on how to properly configure
+Jenkins for secure command-line access.
+
+// link:https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)[CRSF] protection
+
+[source,bash]
+----
+# ssh (Jenkins CLI)
+# JENKINS_SSHD_PORT=[sshd port on master]
+# JENKINS_HOSTNAME=[Jenkins master hostname]
+ssh -p $JENKINS_SSHD_PORT $JENKINS_HOSTNAME declarative-linter < Jenkinsfile
+
+# curl (REST API)
+# Assuming "anonymous read access" has been enabled on your Jenkins instance.
+# JENKINS_URL=[root URL of Jenkins master]
+# JENKINS_CRUMB is needed if your Jenkins master has CRSF protection enabled as it should
+JENKINS_CRUMB=`curl "$JENKINS_URL/crumbIssuer/api/xml?xpath=concat(//crumbRequestField,\":\",//crumb)"`
+curl -X POST -H $JENKINS_CRUMB -F "jenkinsfile=<Jenkinsfile" $JENKINS_URL/pipeline-model-converter/validate
+----
+
+=== Examples
+
+.Invalid Jenkinsfile
+[source,bash]
+----
+# pass a Jenkinsfile that does not contain an "agent" section
+ssh -p 8675 localhost declarative-linter < ./Jenkinsfile
+Errors encountered validating Jenkinsfile:
+WorkflowScript: 25: Not a valid section definition: 'agent'. Some extra configuration is required. @ line 25, column 1.
+   pipeline &#125;
+   ^
+
+WorkflowScript: 25: Missing required section 'agent' @ line 25, column 1.
+   pipeline &#125;
+   ^
+----
+
+.Valid Jenkinsfile
+[source,bash]
+----
+ssh -p 8675 localhost declarative-linter < ./Jenkinsfile
+Jenkinsfile successfully validated.
+----
+
+[[replay]]
+== "Replay" Pipeline Runs with Modifications
+
+Creating Pipelines and debugging failures in Pipelines can be difficult and time consuming.
+The standard development process for Pipeline changes -
```

Comment:
This process is referred to as "the standard" but we never actually describe such a process in the handbook, so I don't think it can/should stand on its own. I understand the desired lead-in but this relies a little too much on pains only a more seasoned Pipeline author (not using Blue Ocean) might have experienced.

What about something like:

> Typically a Pipeline will be defined inside of the classic Jenkins web UI, or by committing to a `Jenkinsfile` in source control. Unfortunately neither approach is ideal for rapid iteration, or prototyping, of a Pipeline. The "Replay" feature allows for quick modifications and execution of an existing Pipeline without changing the Pipeline configuration or creating a new commit. To use the "Replay" feature: